### PR TITLE
DataFrame ADT

### PR DESCRIPTION
**Pull request by psahabu**
### Motivation
The motivation for the DataFrame ADT can be found in the connected issue.

### Limitations
Currently it is not possible to emit a row from a DataFrame because a row requires a variable length, variable type ADT. See #9660 for exploration of this issue.

### Notes
Included in this PR are pandas equivalents to our tests. Note that the only significant differences are:
- slightly different spacing when printing a DataFrame

### CODE REVIEW

File: test/library/standard/DataFrames/DataFrames.chpl [hunk 1]
```diff
@@ -539,15 +616,116 @@ module DataFrames {
...
-          f <~> i + "\t" + d + "\n";
+          f <~> i + "    " + d + "\n";
       }
       f <~> "dtype: " + eltType:string;
     }
+
+    pragma "no doc"
+    proc writeElem(f, i, len: int) {
+      var output = if this.valid(i)
+                   then new string(this[i]: string)
```
Comment (benharsh) [inline test/library/standard/DataFrames/DataFrames.chpl]: Can you add a TODO comment to clean this up once the bug is fixed?

File: test/library/standard/DataFrames/DataFrames.chpl [hunk 2]
```diff
@@ -539,15 +616,116 @@ module DataFrames {
...
+        var n = s.nrows();
+        if n > nMax then nMax = n;
+      }
+      return nMax;
+    }
+
+    proc writeThis(f) {
+      if idx {
+        idx.writeThis(f, this);
+      } else {
```
Comment (benharsh) [inline test/library/standard/DataFrames/DataFrames.chpl]: I feel like we should create an actual Index if the user doesn't provide one, then we can remove this branch.

File: test/library/standard/DataFrames/DataFrames.chpl [hunk 3]
```diff
@@ -539,15 +616,116 @@ module DataFrames {
...
+    var labels: domain(string);
+    var columns: [labels] Series;
+    var idx: Index;
+
+    // TODO: init with labels arg
+
+    proc init(columns: [?D] Series) {
+      this.labels = D;
+      this.columns = columns;
+      this.idx = nil;
```
Comment (benharsh) [inline test/library/standard/DataFrames/DataFrames.chpl]: Isn't a dataframe supposed to take a copy of the given series?

File: test/library/standard/DataFrames/DataFrames.chpl (hunk 2 above)
Comment (psahabu) [reply test/library/standard/DataFrames/DataFrames.chpl]: Right now the practice in the ADTs is that a `nil` idx implies a dense range. #9658 should address this, at which point I agree that we should remove these weird `if idx` branches.

**Comment by psahabu** [comment]
@benharsh mind taking another look? Pretty sure I tackled it all.

**Pull request merged**